CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
```

Comment:
Are the proofs signed, or can anyone submit a proof and it gets gossiped through the network if the proof is verified?

Reply:
I think they wouldn't need to be signed, any proof that is verified should be able to be gossiped, allowing us to decentralize the proof creation. C labs will host an initial "proving service" to do this, but community members should be able to contribute.

This is where I would see issues with proof IDs coming in to play

Reply:
Ya.  Agreed about potential issues with proof IDs.  Seems like it's a DOS attack vector.  E.g. a malicious party could submit many duplicated proof but with different proofIDs to fill up all the full nodes' disks.

Reply:
Yup good points - I plan to key by epoch range like you mentioned @kevjue - specifically `types.PlumoProofEpochs{ firstEpoch, lastEpoch }`

I had originally wanted to follow this path, but thought we might want incremental IDs. But these are good points, and the idea of using a bloom filter makes the incremental IDs unnecessary in my mind

Reply:
Validators should have a fixed list of epoch number pairs (A_1, B_1), (A_2,B_2), ... they accept proofs for. The distance B_i - A_i will be in some fixed set we must determine before carrying out the setup "ceremony." I don't think it makes sense to store (0,32) and (1,33). I think just storing the minimal number of proofs to get from the genesis epoch to the current (or as close as the different fixed lengths let you) is ideal. There's a new circuit @kobigurk has been working on that allows for dummy epoch data and maybe that actually allows for proving an arbitrary number of epochs up to some limit. Hopefully he can chime in here. That would be really nice, as nodes would only have to ceil(t/~120) proofs, where t is the current epoch number.

As for duplicate proofs, nodes should probably just store the first valid proof they receive for a given instance (epoch number pair).